File: sdks/typescript/test/ws-thin-client.test.mjs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025 PlexSpaces Contributors
//
// Unit tests for WsThinClient wire encoding and ThinNodePingResult shape.
//
// These tests cover the wire layer (encode/decode round-trips) without a live
// PlexSpaces server.  Full lifecycle tests (connect → register → ping →
// disconnect → unregistered) are covered by the Rust integration tests in
// crates/node/tests/suite/ws_integration_tests.rs.
//
// Run: node --test test/ws-thin-client.test.mjs

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { pathToFileURL } from 'node:url';
import { fileURLToPath } from 'node:url';
import path from 'node:path';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const DIST = path.resolve(__dirname, '../dist');

const {
  encodeWsFrameNodeRegister,
  encodeWsFrameNodePing,
  encodeWsFrameHeartbeat,
  encodeWsFrameTell,
  encodeWsFrameAsk,
  decodeWsFrame,
} = await import(pathToFileURL(DIST + '/wire/ws-frame-wire.js').href);

// ─── helpers ────────────────────────────────────────────────────────────────

/** Build a minimal synthetic WsFrame binary that carries a NodePingResponse.
 *  Field layout (proto wire):
 *    WsFrame.request_id (field 1, str)
 *    WsFrame.node_ping_response (field 23, len-delimited)
 *      PingResponse.node_id (field 2, str)
 *      PingResponse.sequence_number (field 3, varint)
 *      PingResponse.resources (field 9, len-delimited)
 *        NodeResourceHints.cpu_percent (field 1, fixed32/float)
 *        NodeResourceHints.memory_available_mb (field 2, varint)
 *        NodeResourceHints.available_cores (field 3, varint)
 */
function buildPingResponseFrame({ requestId, nodeId, seqNum, cpuPercent, memoryMb, cores }) {
  // We use the encode helpers we already trust to build the frame bytes.
  // The trick: construct a fake PingResponse bytes manually using the same
  // proto-wire helpers that ws-frame-wire uses internally.
  // Instead, we'll rely on the actual round-trip: encode a NodePing, then
  // build a synthetic response directly via raw bytes (for isolation).

  // Encode a PingResponse manually with raw proto wire encoding.
  const enc = new TextEncoder();
  const parts = [];

  // Field 1 = request_id (string)
  const reqIdBytes = enc.encode(requestId);
  parts.push(...tag(1, 2), ...varint(reqIdBytes.length), ...reqIdBytes);

  // Field 2 = node_id (string)
  const nodeIdBytes = enc.encode(nodeId);
  parts.push(...tag(2, 2), ...varint(nodeIdBytes.length), ...nodeIdBytes);

  // Field 3 = sequence_number (varint)
  parts.push(...tag(3, 0), ...varint(seqNum));

  // Field 9 = resources (NodeResourceHints sub-message)
  const hints = [];
  // Field 1 = cpu_percent (float, fixed32, wire type 5)
  const floatBuf = new Uint8Array(4);
  new DataView(floatBuf.buffer).setFloat32(0, cpuPercent, true);
  hints.push(...tag(1, 5), ...floatBuf);
  // Field 2 = memory_available_mb (uint64 varint)
  hints.push(...tag(2, 0), ...varint(memoryMb));
  // Field 3 = available_cores (uint32 varint)
  hints.push(...tag(3, 0), ...varint(cores));
  const hintsBuf = Uint8Array.from(hints);
  parts.push(...tag(9, 2), ...varint(hintsBuf.length), ...hintsBuf);

  const pingRespBytes = Uint8Array.from(parts);

  // Wrap in WsFrame: field 1 = request_id, field 23 = node_ping_response
  const outerParts = [];
  const outerReqBytes = enc.encode(requestId);
  outerParts.push(...tag(1, 2), ...varint(outerReqBytes.length), ...outerReqBytes);
  outerParts.push(...tag(23, 2), ...varint(pingRespBytes.length), ...pingRespBytes);
  return Uint8Array.from(outerParts);
}

function tag(fieldNum, wireType) {
  return varint((fieldNum << 3) | wireType);
}

function varint(v) {
  const bytes = [];
  let n = BigInt(v);
  do {
    let b = Number(n & 0x7fn);
    n >>= 7n;
    if (n > 0n) b |= 0x80;
    bytes.push(b);
  } while (n > 0n);
  return bytes;
}

// ─── WsFrame encode/decode round-trips ──────────────────────────────────────

describe('ws-frame-wire: NodeRegister encode', () => {
  it('produces a non-empty binary frame', () => {
    const frame = encodeWsFrameNodeRegister(
      'req-1',
      'thin-node-01',
      '',
      { namespace: 'default' },
    );
    assert.ok(frame instanceof Uint8Array);
    assert.ok(frame.length > 0);
  });

  it('decodes back to unknown (server-to-client frame, not decoded by client)', () => {
    // encodeWsFrameNodeRegister is a client→server frame (field 20).
    // decodeWsFrame handles server→client frames only, so this returns 'unknown' — correct.
    const frame = encodeWsFrameNodeRegister('req-1', 'thin-node-01', '', {});
    const decoded = decodeWsFrame(frame);
    assert.equal(decoded.type, 'unknown');
  });
});

describe('ws-frame-wire: NodePing encode', () => {
  it('produces a non-empty binary frame', () => {
    const frame = encodeWsFrameNodePing('ping-req-1', 'thin-node-01', 7);
    assert.ok(frame instanceof Uint8Array);
    assert.ok(frame.length > 0);
  });
});

describe('ws-frame-wire: NodePingResponse decode (with resource hints)', () => {
  it('decodes node_id, sequence_number, cpuPercent, memoryAvailableMb, availableCores', () => {
    const raw = buildPingResponseFrame({
      requestId: 'ping-req-42',
      nodeId:    'server-node-1',
      seqNum:    42,
      cpuPercent: 23.5,
      memoryMb:   4096,
      cores:      8,
    });

    const decoded = decodeWsFrame(raw);
    assert.equal(decoded.type, 'node_ping_response');
    assert.equal(decoded.requestId, 'ping-req-42');
    assert.equal(decoded.nodeId, 'server-node-1');
    // cpuPercent is a float32 — allow small rounding error
    assert.ok(
      Math.abs(decoded.cpuPercent - 23.5) < 0.01,
      `cpuPercent: expected ~23.5, got ${decoded.cpuPercent}`
    );
    assert.equal(decoded.memoryAvailableMb, 4096);
    assert.equal(decoded.availableCores, 8);
  });

  it('returns zeros for missing resource hints', () => {
    // Build a PingResponse with no resources sub-message
    const enc = new TextEncoder();
    const nodeId = enc.encode('node-x');
    const pingParts = [];
    pingParts.push(...tag(2, 2), ...varint(nodeId.length), ...nodeId);
    const pingBytes = Uint8Array.from(pingParts);
    const outerReqBytes = enc.encode('ping-no-res');
    const outer = [
      ...tag(1, 2), ...varint(outerReqBytes.length), ...outerReqBytes,
      ...tag(23, 2), ...varint(pingBytes.length), ...pingBytes,
    ];
    const decoded = decodeWsFrame(Uint8Array.from(outer));
    assert.equal(decoded.type, 'node_ping_response');
    assert.equal(decoded.cpuPercent, 0);
    assert.equal(decoded.memoryAvailableMb, 0);
    assert.equal(decoded.availableCores, 0);
  });
});

describe('ws-frame-wire: Heartbeat encode', () => {
  it('produces a non-empty binary frame', () => {
    const frame = encodeWsFrameHeartbeat('hb-1', 'thin-node-01');
    assert.ok(frame instanceof Uint8Array);
    assert.ok(frame.length > 0);
  });
});

describe('ws-frame-wire: Tell encode — namespace extracted from canonical actor ID', () => {
  it('encodes namespace from canonical actor ID', () => {
    const actorId = 'alice//ChatClient::lobby@server-node-1';
    const payload = new TextEncoder().encode(JSON.stringify({ text: 'hello' }));
    const frame = encodeWsFrameTell('tell-1', actorId, 'send', payload);
    assert.ok(frame instanceof Uint8Array);
    assert.ok(frame.length > 0);
    // The frame is a client→server tell; decodeWsFrame won't decode it (field 10 is incoming tell from server).
    // Just verify it's non-empty and correct type won't crash decode.
    const decoded = decodeWsFrame(frame);
    // Field 10 as a client-to-server tell is decoded as incoming_tell when server sends it back.
    // When client sends it, it's decoded as incoming_tell (same field, same direction for testing).
    // Accept both 'incoming_tell' and 'unknown' since direction is client→server.
    assert.ok(['incoming_tell', 'unknown'].includes(decoded.type));
  });
});

describe('ws-frame-wire: Ask encode', () => {
  it('encodes namespace from canonical actor ID and includes timeout', () => {
    const actorId = 'room//ChatRoomActor::lobby@server-node-1';
    const payload = new TextEncoder().encode(JSON.stringify({ action: 'join' }));
    const frame = encodeWsFrameAsk('ask-1', actorId, 'join', payload, 5000);
    assert.ok(frame instanceof Uint8Array);
    assert.ok(frame.length > 0);
  });
});

// ─── ThinNodePingResult shape (interface contract) ──────────────────────────

describe('ThinNodePingResult interface fields', () => {
  it('decoded node_ping_response matches ThinNodePingResult fields', () => {
    const raw = buildPingResponseFrame({
      requestId: 'iface-check',
      nodeId:    'full-node-99',
      seqNum:    1,
      cpuPercent: 55.0,
      memoryMb:   8192,
      cores:      16,
    });
    const frame = decodeWsFrame(raw);
    assert.equal(frame.type, 'node_ping_response');

    // Verify all ThinNodePingResult fields are present and typed correctly
    const result = {
      success: true,
      nodeId:            frame.nodeId,
      cpuPercent:        frame.cpuPercent,
      memoryAvailableMb: frame.memoryAvailableMb,
      availableCores:    frame.availableCores,
    };
    assert.equal(typeof result.success,            'boolean');
    assert.equal(typeof result.nodeId,             'string');
    assert.equal(typeof result.cpuPercent,         'number');
    assert.equal(typeof result.memoryAvailableMb,  'number');
    assert.equal(typeof result.availableCores,     'number');
    assert.equal(result.nodeId, 'full-node-99');
    assert.ok(Math.abs(result.cpuPercent - 55.0) < 0.01);
    assert.equal(result.memoryAvailableMb, 8192);
    assert.equal(result.availableCores, 16);
  });
});

File: sdks/go/plexspaces/ws_thin_client_test.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025 PlexSpaces Contributors
//
// Unit tests for WsThinClient types and interface contracts.
//
// Full lifecycle tests (connect → register → ping → disconnect → unregistered)
// require a live server and are covered by the Rust integration tests in
// crates/node/tests/suite/ws_integration_tests.rs.

package plexspaces_test

import (
	"context"
	"testing"

	"github.com/bhatti/PlexSpaces/sdks/go/plexspaces"
)

// TestWsThinClientOptions verifies the options struct can be constructed with
// all fields populated — guards against accidental field removal.
func TestWsThinClientOptions_Fields(t *testing.T) {
	opts := plexspaces.WsThinClientOptions{
		WsURL:     "ws://localhost:8091/ws",
		JwtToken:  "test.jwt.token",
		NodeID:    "thin-node-go-01",
		Tenant:    "default",
		Namespace: "test-ns",
	}
	if opts.WsURL == "" {
		t.Error("WsURL must not be empty")
	}
	if opts.Tenant == "" {
		t.Error("Tenant must not be empty")
	}
}

// TestThinNodePingResult_Fields verifies all resource hint fields are present
// and correctly typed on ThinNodePingResult.
func TestThinNodePingResult_Fields(t *testing.T) {
	result := plexspaces.ThinNodePingResult{
		NodeID:            "server-node-1",
		CPUPercent:        23.5,
		MemoryAvailableMB: 4096,
		AvailableCores:    8,
	}
	if result.NodeID != "server-node-1" {
		t.Errorf("NodeID = %q, want server-node-1", result.NodeID)
	}
	if result.CPUPercent != 23.5 {
		t.Errorf("CPUPercent = %v, want 23.5", result.CPUPercent)
	}
	if result.MemoryAvailableMB != 4096 {
		t.Errorf("MemoryAvailableMB = %v, want 4096", result.MemoryAvailableMB)
	}
	if result.AvailableCores != 8 {
		t.Errorf("AvailableCores = %v, want 8", result.AvailableCores)
	}
}

// TestNewWsThinClient verifies the constructor returns a non-nil client.
func TestNewWsThinClient_NonNil(t *testing.T) {
	c := plexspaces.NewWsThinClient(plexspaces.WsThinClientOptions{
		WsURL: "ws://localhost:8091/ws",
	})
	if c == nil {
		t.Fatal("NewWsThinClient must return non-nil client")
	}
}

// TestWsThinClient_Connect_ReturnsError verifies the stub returns an error
// (not implemented) rather than panicking or hanging.
func TestWsThinClient_Connect_ReturnsError(t *testing.T) {
	c := plexspaces.NewWsThinClient(plexspaces.WsThinClientOptions{
		WsURL: "ws://127.0.0.1:0/ws",
	})
	_, err := c.Connect(context.Background())
	if err == nil {
		t.Fatal("Connect on stub must return an error")
	}
}

// TestWsThinClient_Tell_ReturnsError verifies the stub returns an error.
func TestWsThinClient_Tell_ReturnsError(t *testing.T) {
	c := plexspaces.NewWsThinClient(plexspaces.WsThinClientOptions{})
	err := c.Tell("actor//T::ns@node", "ping", nil)
	if err == nil {
		t.Fatal("Tell on stub must return an error")
	}
}

// TestWsThinClient_Ask_ReturnsError verifies the stub returns an error.
func TestWsThinClient_Ask_ReturnsError(t *testing.T) {
	c := plexspaces.NewWsThinClient(plexspaces.WsThinClientOptions{})
	_, err := c.Ask(context.Background(), "actor//T::ns@node", "get", nil)
	if err == nil {
		t.Fatal("Ask on stub must return an error")
	}
}

// TestWsThinClient_PingNode_ReturnsError verifies the stub returns an error.
func TestWsThinClient_PingNode_ReturnsError(t *testing.T) {
	c := plexspaces.NewWsThinClient(plexspaces.WsThinClientOptions{})
	_, err := c.PingNode(context.Background(), "target-node")
	if err == nil {
		t.Fatal("PingNode on stub must return an error")
	}
}

// TestWsThinClient_NodeID_Empty verifies NodeID returns empty on stub.
func TestWsThinClient_NodeID_Empty(t *testing.T) {
	c := plexspaces.NewWsThinClient(plexspaces.WsThinClientOptions{})
	if id := c.NodeID(); id != "" {
		t.Errorf("NodeID on stub must return empty, got %q", id)
	}
}
